contracts/rateproducer/rateproducer.cpp

```diff
@@ -4,257 +4,470 @@
+      double              last_vote_weight = 0;
+      double              proxied_vote_weight= 0;
+      bool                is_proxy = 0;
+      uint32_t            flags1 = 0;
+      uint32_t            reserved2 = 0;
+      eosio::asset        reserved3;
 
-      //TODO: bp must be a registered block producer
+      uint64_t primary_key()const { return owner.value; }
 
-      // the payload must be ratings_json.
-      check(ratings_json[0] == '{', "payload must be ratings_json");
-      check(ratings_json[ratings_json.size()-1] == '}', "payload must be ratings_json");
+      EOSLIB_SERIALIZE( voter_info, (owner)(proxy)(producers)(staked)(last_vote_weight)(proxied_vote_weight)(is_proxy)(flags1)(reserved2)(reserved3) )
+   };
 
-      // upsert bp rating
-      producers_table bps(_self, _self.value);
-      auto uniq_rating = (static_cast<uint128_t>(user.value) << 64) | bp.value;
+    typedef eosio::multi_index< "voters"_n, voter_info >  voters_table; 
 
-      auto uniq_rating_index = bps.get_index<name("uniqrating")>();
-      auto existing_rating = uniq_rating_index.find(uniq_rating);
+   int get_voters (name name) {
+      voters_table _voters(system_account, system_account.value);
+      auto it = _voters.find(name.value);
+      if(it==_voters.end()){
+          return 0;
+      }
+      return it->producers.size();
+   }
+
+     eosio::name get_proxy ( eosio::name name) {
+
+      voters_table _voters(system_account, system_account.value);
+      auto it = _voters.find(name.value);
+      if(it==_voters.end()){
+          eosio::name result("");
+          return result;
+      }
+      return it->proxy;
+   }
 
-      uint64_t now = eosio::current_time_point().time_since_epoch().count();
+} /// namespace eosio
 
-      if( existing_rating == uniq_rating_index.end() ) {
+CONTRACT rateproducer : public contract {
+  public:
+    using contract::contract;
+      
+      ACTION rate(name user, 
+                  name bp, 
+                  int8_t transparency,
+                  int8_t infrastructure,
+                  int8_t trustiness,
+                  int8_t community,
+                  int8_t development) {
+        check( (transparency+infrastructure+trustiness+community+development), "Error vote must have value for at least one category");  
+        check( (MINVAL<= transparency &&  transparency<=MAXVAL ), "Error transparency value out of range");
+        check( (MINVAL<= infrastructure &&  infrastructure<=MAXVAL ), "Error infrastructure value out of range" );
+        check( (MINVAL<= trustiness &&  trustiness<=MAXVAL ), "Error trustiness value out of range" );
+        check( (MINVAL<= development &&  development <=MAXVAL ), "Error development value out of range" );
+        check( (MINVAL<= community &&  community<=MAXVAL ), "Error community value out of range" );
+      
+         //checks if the bp is active 
+        check(is_blockproducer(bp),"votes are allowed only for registered block producers");
```

Comment:
@kriskoin
This is a very important validation , thanks for adding it!  It does however create some complications since we are currently running our contracts on the jungle testnet and sourcing our block producers form the EOS Mainnnet.
 
@tetogomez
There are no changes required in the contract but the frontend must now source all block producers from a jungle endpoint.

 You can use our own endpoint  https://jungle.eosio.cr , this may be just a simple change in environmental vars.

Reply:
@xavier506 , should i disable it ? maybe temporary?

Reply:
@kriskoin  No need to disable it , we can work around it , we will configure the app to use jungle producers for the staging environment and mainnet producers for production.